news/_posts/2020-11-30-from-legacy-to-future-architecture-connecting-the-dots.md

```diff
@@ -0,0 +1,308 @@
+
+For example, when an order's carrier is updated, an email should be sent to the customer informing them that the order is on its way. This is done automatically when the action is performed using the appropriate service. However, if the order's carrier is modified manually through ObjectModel, for example by a module developer not aware of this rule, then the required email is not sent.
+
+This problem of thick service layers manipulating thin models devoid of domain logic was identified over 15 years ago by Martin Fowler, who described it the [Anemic Domain Model][anemic-domain-model] anti-pattern. His answer was simple: business logic should be put in the same place as data is – [entity objects][ddd-entities].
+
+This means that **there should be one single, obvious way to perform any given system action**. If there is only one possible way to update an order's carrier, and if this is done in an intuitive way through the Order entity, then we can ensure that all business rules are being enforced regardless of where this task is initiated.
+
+Here is how you would update an order's shipping details:
+
+```php
+$someCustomer // this is a Customer
+  ->getOrderById($orderId) // this returns an Order
+	->updateShippingDetails($newCarrier, $trackingNumber);
+```
+
+Of course, this is easier said that done: getting to this kind of workflow in PrestaShop would require a complete rewrite of the Core, and as we said earlier, we don't want to do that. Which brings us to the next issue...
+
+### Extensions need a reliable interface to the Core
+
+As the Core evolves, its components will inevitably need to be split, renamed, change their methods' signatures, and so on. However, doing this would provoke breaking changes for extensions that rely on the core components' interfaces staying unchanged.
+
+This situation can be assimilated to the [double bind dilemma][double-bind-dilemma], where two conflicting needs prevent the situation from evolving satisfactorily:
+
+- The system is complex, has bugs or needs to accommodate new features, **so components must change**.
+- Extensions rely on the current implementation, **so component mustn't change**.
+
+This has been slowing down PrestaShop's progress for quite a while. But luckily, I think there's a way out. 
+
+Let's revisit the D in [SOLID][solid] for a minute. The [Dependency Inversion Principle][dependency-inversion-principle] states the following:
+
+> 1. High level modules should not depend on low level modules. Both must depend on abstractions.
+> 2. Abstractions should not depend on details. The details must depend on the abstractions.
```

Comment:
```suggestion
> 2. Abstractions should not depend on details. The details must depend on abstractions.
```

Reply:
Actually I think that it sounds better without the first "The"
```suggestion
> 2. Abstractions should not depend on details. Details must depend on abstractions.
```